boa_engine/src/environments/runtime.rs

```diff
@@ -852,3 +843,323 @@ impl BindingLocator {
+            let env = self
+                .realm
+                .environments
+                .stack
+                .get(env_index)
+                .expect("environment index must be in range");
+
+            match env {
+                Environment::Declarative(env) => {
+                    if env.poisoned.get() {
+                        let compile = env.compile.borrow();
+                        if compile.is_function() {
+                            if let Some(b) = compile.get_binding(name) {
+                                let mut bindings = self
+                                    .realm
+                                    .environments
+                                    .stack
+                                    .get(b.environment_index)
+                                    .expect("environment index must be in range")
+                                    .as_declarative()
+                                    .expect("environment must be declarative")
+                                    .bindings
+                                    .borrow_mut();
+                                let binding = bindings
+                                    .get_mut(b.binding_index)
+                                    .expect("binding index must be in range");
+                                *binding = Some(value.clone());
+                                return Ok(true);
+                            }
+                        }
+                    } else if !env.with.get() {
+                        return Ok(false);
+                    }
+                }
+                Environment::Object(o) => {
+                    let o = o.clone();
+                    let key: JsString = self
+                        .interner()
+                        .resolve_expect(name.sym())
+                        .into_common(false);
+                    if o.has_property(key.clone(), self)? {
+                        if let Some(unscopables) = o.get(JsSymbol::unscopables(), self)?.as_object()
+                        {
+                            if unscopables.get(key.clone(), self)?.to_boolean() {
+                                continue;
+                            }
+                        }
+                        return o.set(key, value.clone(), true, self);
+                    }
+                }
+            }
+        }
+
+        Ok(false)
+    }
+
+    /// Delete a binding form an object environment if it exists.
+    ///
+    /// Returns a tuple of `(found, deleted)`.
+    pub(crate) fn delete_binding_from_objet_environment(
+        &mut self,
+        name: Identifier,
+    ) -> JsResult<(bool, bool)> {
+        for env_index in (0..self.realm.environments.stack.len()).rev() {
+            let env = self
+                .realm
+                .environments
+                .stack
+                .get(env_index)
+                .expect("environment index must be in range");
```

Comment:
```suggestion
            let env = self.environment_expect(env_index);
```
Maybe it would be good to refactor into a utility function, this pattern seems to be in many functions